### PR TITLE
Copy all of ``OptionsManagerMixIn`` methods into ``_ArgumentsManager``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Sorry for the size of this PR. Basically these are all the methods defined on `OptionsManagerMixIn`. Sadly they are all public so they will need to be deprecated.
I thought it would be better to move them on `_ArgumentsManager` and deprecate from there. This would allow us to deprecate instantiating `OptionsManagerMixIn` and completely disregard that file until `3.0` gets released.

### CODE REVIEW

File: tests/lint/unittest_lint.py [hunk 1]
```diff
@@ -710,7 +710,7 @@ def test_pylintrc_parentdir_no_package() -> None:
                     assert config.find_pylintrc() == expected
 
 
-class _CustomPyLinter(PyLinter):  # pylint: disable=too-many-ancestors
+class _CustomPyLinter(PyLinter):
```
Comment (Pierre-Sassoulas) [inline tests/lint/unittest_lint.py:713]: 🙏

**Comment by DanielNoord** [comment]
The coverage is the same as for `OptionsManagerMixIn`. Will probably be fixed when I start working on deprecation warnings here.

File: pylint/config/arguments_manager.py [hunk 2]
```diff
@@ -161,3 +225,274 @@ def _parse_command_line_configuration(
...
+                        sect = sect[len("pylint.") :]
+                    if not sect.isupper() and values:
+                        parser._sections[sect.upper()] = values
+
+        if not verbose:
+            return
+        if config_file and config_file.exists():
+            msg = f"Using config file '{config_file}'"
+        else:
+            msg = "No config file found, using default configuration"
```
Comment (Pierre-Sassoulas) [inline pylint/config/arguments_manager.py:420]: This will never be executed because of line 396.

**Review by Pierre-Sassoulas — COMMENTED** [review]
There's a a whole untested function "_expand_default" I think we should cover it or remove it unless it's optparse specific code. Also I think we could replace OptionManagerMixin by ArgumentManger like this:

```python

class OptionManagerMixin:

   def __init__(...)
        warnings.warn(...)
       self.argument_manager = ...
       
    def old_option_manager_function(...)
        return self.argument_manager.new_function(...)
```
To avoid duplication.

File: pylint/config/arguments_manager.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/config/arguments_manager.py:420]: I think it can be executed if L393 fails?

**Comment by DanielNoord** [comment]
> There's a a whole untested function "_expand_default" I think we should cover it or remove it unless it's optparse specific code. 

It's `optparse` specific. It's used in `_patch_optparse`.

> Also I think we could replace OptionManagerMixin by ArgumentManger like this:
> 
> ```python
> class OptionManagerMixin:
> 
>    def __init__(...)
>         warnings.warn(...)
>        self.argument_manager = ...
>        
>     def old_option_manager_function(...)
>         return self.argument_manager.new_function(...)
> ```
> 
> To avoid duplication.

I'd rather have the duplication for a little bit. Imo, it's much easier to just deprecate `OptionManagerMixin` and the complete file and work within the `ArgumentsManager` file. We need to keep instantiating `OptionManagerMixin` while `pyreverse` and `ConfigurationMixIn` are not yet on `argparse`. With the method proposed in this PR that work doesn't slow us down while working on refactoring these methods. We can start refactoring them already without impacting `pyreverse`.

_In general I prefer cutting ties with anything in the `optparse` files as strictly as possible. I think it makes it easier to see how far along in the process we are._

Obviously if you have a strong preference for doing it within the `OptionManagerMixin` class I will change this around!

File: pylint/config/arguments_manager.py [hunk 3]
```diff
@@ -161,3 +225,274 @@ def _parse_command_line_configuration(
...
+        """Initialize configuration using default values."""
+        for provider in self.options_providers:
+            provider.load_defaults()
+
+    def read_config_file(
+        self, config_file: Optional[Path] = None, verbose: bool = False
+    ) -> None:
+        """Read the configuration file but do not load it (i.e. dispatching
+        values to each option's provider)
+        """
```
Comment (Pierre-Sassoulas) [inline pylint/config/arguments_manager.py:392]:
```suggestion
        """
        if not config_file:
            if verbose:
                print("No config file found, using default configuration", file=sys.stderr)
            return
```

File: pylint/config/arguments_manager.py [hunk 4]
```diff
@@ -161,3 +225,274 @@ def _parse_command_line_configuration(
...
+                    if not sect.isupper() and values:
+                        parser._sections[sect.upper()] = values
+
+        if not verbose:
+            return
+        if config_file and config_file.exists():
+            msg = f"Using config file '{config_file}'"
+        else:
+            msg = "No config file found, using default configuration"
+        print(msg, file=sys.stderr)
```
Comment (Pierre-Sassoulas) [inline pylint/config/arguments_manager.py:421]:
```suggestion
        if verbose:
             print(f"Using config file '{config_file}'", file=sys.stderr)     
```

File: pylint/config/arguments_manager.py (hunk 2 above)
Comment (Pierre-Sassoulas) [reply pylint/config/arguments_manager.py:420]: You're right.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: pylint/config/arguments_manager.py [hunk 5]
```diff
@@ -161,3 +225,274 @@ def _parse_command_line_configuration(
...
+            else:
+                # Use this encoding in order to strip the BOM marker, if any.
+                with open(config_file, encoding="utf_8_sig") as fp:
+                    parser.read_file(fp)
+                # normalize each section's title
+                for sect, values in list(parser._sections.items()):
+                    if sect.startswith("pylint."):
+                        sect = sect[len("pylint.") :]
+                    if not sect.isupper() and values:
+                        parser._sections[sect.upper()] = values
```
Comment (Pierre-Sassoulas) [inline pylint/config/arguments_manager.py:413]:
```suggestion
        config_file = Path(os.path.expandvars(config_file)).expanduser()
        if not config_file.exists():
            raise OSError(f"The config file {str(config_file)} doesn't exist!")
        parser = self.cfgfile_parser
        if config_file.suffix == ".toml":
            try:
                self._parse_toml(config_file, parser)
            except tomllib.TOMLDecodeError:
                pass
        else:
            # Use this encoding in order to strip the BOM marker, if any.
            with open(config_file, encoding="utf_8_sig") as fp:
                parser.read_file(fp)
            # normalize each section's title
            for sect, values in list(parser._sections.items()):
                if sect.startswith("pylint."):
                    sect = sect[len("pylint.") :]
                if not sect.isupper() and values:
                    parser._sections[sect.upper()] = values
```

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**